Testing matrix (#256)
plone/meta: UPGRADE.md

# Upgrade guide

This file describes the breaking changes between major versions.

## From `main` to `2.x`

### Test matrix

You can now define the combinations of Plone versions and Python versions to be tested.

`plone.meta` provides a default combination (the same as used by Plone itself), but can be overridden in the file `.meta.toml`.

```toml
[tox]
test_matrix = {"6.2" = ["3.14", "3.11"], "6.1" = ["3.13", "3.12", "3.11", "3.10", "3.9"]}
```

This will be used to generate the necessary `tox` environments and the GitHub Actions that will be run on each pull request.

> [!TIP]

> > 🍀 `plone.meta` tries to be a bit more environmental friendly.

Reviewer: ```suggestion
> 🍀 `plone.meta` tries to be a bit more environmentally friendly.
```

> On GitHub, only the first and last Python versions will be added for testing.

### Constraints

The `constraints_file` option in `.meta.toml`'s `[tox]` table was renamed to `constraints_files`, and the type of its value was changed from a string to a dictionary.

This option continues to be optional.

The dictionary keys must be Plone versions, and each key's value must be the constraints file for that Plone version.

```toml
[tox]
# OLD
constraints_file = "https://example.org/my-custom-constraints.txt"
# NEW
constraints_files = {"6.2" = "https://example.org/constraints.6.2.txt", "6.1" = "https://example.org/constraints.6.1.txt"}
```

### GitHub Actions

The `py_versions` option in `.meta.toml`'s `[github]` table is deprecated.

> We use the new `test_matrix` option from `[tox]` table as we now can run multiple python versions from within `tox` itself.

Reviewer: ```suggestion
We use the new `test_matrix` option from `[tox]` table, as we now can run multiple Python versions from within `tox` itself.
```

# Upgrade guide

> Here are listed the breaking changes between major versions.

Reviewer: Do you want to document only breaking changes, or also include feature additions between minor versions and how to use them? If you want to include the latter, then I'd append ` and feature additions between minor versions`.

```suggestion
This file describes the breaking changes between major versions.
```

## From `main` to `2.x`

### Test matrix

You can now define the combinations of Plone versions and Python versions to be tested.

> `plone.meta` provides a default (the same as used by Plone itself),

Reviewer: ```suggestion
`plone.meta` provides a default combination (the same as used by Plone itself), but can be overridden in the file `.meta.toml`.
```

> but can be overridden on `.meta.toml`.

Reviewer: ```suggestion
```

```toml
[tox]
test_matrix = {"6.2" = ["3.14", "3.11"], "6.1" = ["3.13", "3.12", "3.11", "3.10", "3.9"]}
```

This will be used to generate the necessary `tox` environments
and the GitHub Actions that will be run on each pull request.

> [!TIP]
> 🍀 To be a bit more environmental friendly, on GitHub,
> only the first and last Python versions will be added for testing.

### Constraints

> The `constraints_file` option on `.meta.toml`'s `[tox]` table got renamed to `constraints_files`

Reviewer: ```suggestion
The `constraints_file` option in `.meta.toml`'s `[tox]` table was renamed to `constraints_files`, and the type of its value was changed from a string to a dictionary.
```

and its value change from a string, to a dictionary.

This option continues to be optional.

> The dictionary keys needs to be plone versions

Reviewer: ```suggestion
The dictionary keys must be Plone versions, and each key's value must be the constraints file for that Plone version.
```

and the value the constraints file for that Plone version.

```toml
[tox]
# OLD
constraints_file = "https://example.org/my-custom-constraints.txt"
# NEW
constraints_files = {"6.2" = "https://example.org/constraints.6.2.txt", "6.1" = "https://example.org/constraints.6.1.txt"}
```

### GitHub Actions

> The `py_versions` option on `.meta.toml`'s `[github]` table is deprecated,

Reviewer: ```suggestion
The `py_versions` option in `.meta.toml`'s `[github]` table is deprecated.
```

> we use the new `test_matrix` option from `[tox]` table

Reviewer: ```suggestion
Use the new `test_matrix` option from the `[tox]` table, as you can now run multiple Python versions from within `tox` itself.
```

as we now can run multiple python versions from within `tox` itself.

### GitHub variables

> The GitHub variables `TEST_OS_VERSIONS` and `TEST_PYTHON_VERSIONS` are deprecated

Reviewer: ```suggestion
The GitHub variables `TEST_OS_VERSIONS` and `TEST_PYTHON_VERSIONS` are deprecated and no longer used.
```

and no longer being used.

### GitLab images

> The `custom_image` option on `.meta.toml`'s `[gitlab]` table got renamed to `custom_images`

Reviewer: ```suggestion
The `custom_image` option in `.meta.toml`'s `[gitlab]` table was renamed to `custom_images`, and its value type was changed from a string to a dictionary.
```

and its value change from a string, to a dictionary.

This option continues to be optional.

> The dictionary keys needs to be python versions

Reviewer: ```suggestion
The dictionary keys must be Python versions, and the values a Docker image for that Python version.
```
